tokio-sync/src/semaphore.rs

```diff
@@ -0,0 +1,1116 @@
+    fn has_available_permits(&self) -> bool {
+        self.0 & NUM_FLAG == NUM_FLAG
+    }
+
+    fn has_waiter(&self, stub: &WaiterNode) -> bool {
+        !self.has_available_permits() && !self.is_stub(stub)
+    }
+
+    /// Try to acquire a permit
+    ///
+    /// # Return
+    ///
+    /// Returns `true` if the permit was acquired, `false` otherwise. If `false`
+    /// is returned, it can be assumed that `State` represents the head pointer
+    /// in the mpsc channel.
+    fn acquire_permit(&mut self, stub: &WaiterNode) -> bool {
+        if !self.has_available_permits() {
+            return false;
+        }
+
+        debug_assert!(self.waiter().is_none());
+
+        self.0 -= 1 << NUM_SHIFT;
+
+        if self.0 == NUM_FLAG {
+            // Set the state to the stub pointer.
+            self.0 = stub as *const _ as usize;
+        }
+
+        true
+    }
+
+    /// Release permits
+    ///
+    /// Returns `true` if the permits were accepted.
+    fn release_permits(&mut self, permits: usize, stub: &WaiterNode) {
+        debug_assert!(permits > 0);
+
+        if self.is_stub(stub) {
+            self.0 = (permits << NUM_SHIFT) | NUM_FLAG | (self.0 & CLOSED_FLAG);
+            return;
+        }
+
+        debug_assert!(self.has_available_permits());
+
+        self.0 += permits << NUM_SHIFT;
+    }
+
+    fn is_waiter(&self) -> bool {
+        self.0 & NUM_FLAG == 0
+    }
+
+    /// Returns the waiter, if one is set.
+    fn waiter(&self) -> Option<NonNull<WaiterNode>> {
+        if self.is_waiter() {
+            let waiter = NonNull::new(self.as_ptr())
+                .expect("null pointer stored");
+
+            Some(waiter)
+        } else {
+            None
+        }
+    }
+
+    /// Assumes `self` represents a pointer
+    fn as_ptr(&self) -> *mut WaiterNode {
+        (self.0 & !CLOSED_FLAG) as *mut WaiterNode
+    }
+
+    /// Set to a pointer to a waiter.
+    ///
+    /// This can only be done from the full state.
+    fn set_waiter(&mut self, waiter: NonNull<WaiterNode>) {
+        let waiter = waiter.as_ptr() as usize;
+        debug_assert!(waiter & NUM_FLAG == 0);
+        debug_assert!(!self.is_closed());
+
+        self.0 = waiter;
+    }
+
+    fn is_stub(&self, stub: &WaiterNode) -> bool {
+        self.as_ptr() as usize == stub as *const _ as usize
+    }
+
+    /// Load the state from an AtomicUsize.
+    fn load(cell: &AtomicUsize, ordering: Ordering) -> SemState {
+        let value = cell.load(ordering);
+        debug!(" + SemState::load; value = {}", value);
+        SemState(value)
+    }
+
+    /// Swap the values
+    fn swap(&self, cell: &AtomicUsize, ordering: Ordering) -> SemState {
+        let prev = SemState(cell.swap(self.to_usize(), ordering));
+        debug_assert_eq!(prev.is_closed(), self.is_closed());
+        prev
+    }
+
+    /// Compare and exchange the current value into the provided cell
+    fn compare_exchange(&self,
```

Comment:
What do you think about using [`Atomic<SemState>`](https://docs.rs/atomic/0.4.3/atomic/struct.Atomic.html) instead? It might simplify things quite a bit...

Reply:
Looking into `Atomic`, the impl is pretty intense... (though clever).

The bigger issue is that I don't think it supports all the cases we need. There are some RMW ops like `fetch_or` being used, which I don't think `Atomic` supports.